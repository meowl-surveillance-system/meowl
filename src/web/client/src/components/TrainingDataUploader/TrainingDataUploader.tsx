import React, { Component, ChangeEvent } from "react";
import { Button, Container, Paper, TableContainer, Table, TableHead, TableRow, TableCell, TableBody } from "@material-ui/core";

export interface Props {
    isLoggedIn: boolean;
    onAuthChange: (authState: boolean) => void;
}

export interface State {
    selectedFiles: File[]
}

export default class TrainingDataUploader extends Component<Props, State> {
    fileInputRef: HTMLInputElement | null;

    constructor(props: Props) {
        super(props);
        this.fileInputRef = null;
        this.state = {
            selectedFiles: []
        };
    }

    onFilesChange(event: ChangeEvent<HTMLInputElement>) {
        event.persist();
        if (event.target && event.target.files) {
            this.setState({ selectedFiles: Array.from(event.target.files) });
        } else {
            this.setState({ selectedFiles: [] });
        }
    }

    onFilesUpload() {
        this.state.selectedFiles.forEach(async (file) => {
            return await this.uploadFile(file);
        });
        this.setState({ selectedFiles: [] });
    }

    async uploadFile(file: File): Promise<boolean> {
        const formData = new FormData();
        formData.append('TrainingData', file, file.name);
        return await fetch('/cv/upload/trainingData', {
            method: 'PUT',
            mode: 'same-origin',
            credentials: 'same-origin',
            body: formData,
        }).then((response) => {
            if (response.ok) {
                return true;
            } else {
                console.error(response.json());
                return false;
            }
        }).catch((error) => {
            console.error(error);
            return false;
        });
    }

    promptFileInput() {
        if (this.fileInputRef) {
            this.fileInputRef.click();
        } else {
            console.error('Unable to click on file input element.')
        }
    }

    render(): JSX.Element {
        return (
            <Container component="main" maxWidth="xs">
                <TableContainer component={Paper}>
                    <Table className='trainingDataFile' aria-label="simple table">
                        <TableHead>
                            <TableRow>
                                <TableCell align="right">File Name</TableCell>
                                <TableCell align="right">Size</TableCell>
                                <TableCell align="right">Last Modified</TableCell>
                            </TableRow>
                        </TableHead>
                        <TableBody>
                            {this.state.selectedFiles.map((file) => (
                                <TableRow key={file.name}>
                                    <TableCell component="th" scope="row">
                                        {file.name}
                                    </TableCell>
                                    <TableCell align="right">{file.size}</TableCell>
                                    <TableCell align="right">{new Date(file.lastModified).toDateString()}</TableCell>
                                </TableRow>
                            ))}
                        </TableBody>
                    </Table>
                </TableContainer>
                <input
                    accept="image/*,.pdf,.doc,.docx,.xls,.xlsx"
                    id="file-input"
                    ref={(ref) => { this.fileInputRef = ref }}
                    multiple={true}
                    style={{ display: 'none' }}
                    type="file"
                    onChange={(e) => this.onFilesChange(e)}
                />
                <Button onClick={() => this.promptFileInput()}>
                    Select
                    </Button>
                <Button onClick={() => this.onFilesUpload()}>
                    Upload
                    </Button>
            </Container>
        );
    }
}
