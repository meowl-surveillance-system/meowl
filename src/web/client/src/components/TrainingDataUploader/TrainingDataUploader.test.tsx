import React from "react";
import { mount, ShallowWrapper, shallow } from "enzyme";
import { render } from "@testing-library/react";
import { State } from "./TrainingDataUploader";
import TrainingDataUploader from "./TrainingDataUploader";

describe("TrainingDataUploader component", () => {
  let trainingDataUploaderWrapper: ShallowWrapper<{}, State, TrainingDataUploader>;
  let onAuthChangeMock: jest.Mock;
  beforeEach(() => {
    onAuthChangeMock = jest.fn();
    trainingDataUploaderWrapper = shallow(
      <TrainingDataUploader
        isLoggedIn={false}
        onAuthChange={onAuthChangeMock}
      />
    );
    // console.log(trainingDataUploaderWrapper.debug());
  });

  it("should render the TrainingDataUploader component", () => {
    const { container } = render(
      <TrainingDataUploader
        isLoggedIn={false}
        onAuthChange={onAuthChangeMock}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call uploadFile twice', () => {
    const fileStubs: File[] = [
      new File([], 'dummy.txt'),
      new File([], 'dummy2.txt'),
    ];
    // const uploadFileSpy: jest.SpyInstance = jest.spyOn(TrainingDataUploader, '');
    trainingDataUploaderWrapper = trainingDataUploaderWrapper.setState({ selectedFiles: fileStubs });
    trainingDataUploaderWrapper.instance().onFilesUpload();
  });
});
