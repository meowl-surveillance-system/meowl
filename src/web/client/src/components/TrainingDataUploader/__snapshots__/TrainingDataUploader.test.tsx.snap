// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrainingDataUploader component should render the TrainingDataUploader component 1`] = `
<main
  class="MuiContainer-root MuiContainer-maxWidthXs"
>
  <div
    class="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      aria-label="simple table"
      class="MuiTable-root trainingDataFile"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          >
            File Name
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          >
            Size
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          >
            Last Modified
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      />
    </table>
  </div>
  <input
    accept="image/*,.pdf,.doc,.docx,.xls,.xlsx"
    id="file-input"
    multiple=""
    style="display: none;"
    type="file"
  />
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Select
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Upload
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</main>
`;
